streamer/controller_node.py

```diff
@@ -196,44 +243,55 @@ def start(self, output_dir: str,
 
         outputs.append(TextOutputStream(text_pipe, input))
 
-    self._nodes.append(TranscoderNode(input_config,
-                                      pipeline_config,
+    self._nodes.append(TranscoderNode(inputs,
+                                      self._pipeline_config,
                                       outputs))
 
-    self._nodes.append(PackagerNode(pipeline_config,
+    output_dir = self._output_dir
+    
+    # If the inputs list was a period in multiperiod_inputs_list, create a nested directory
+    # and put that period in it.
+    if period_dir:
+      output_dir = os.path.join(output_dir, period_dir)
+      os.mkdir(output_dir)
+
+    self._nodes.append(PackagerNode(self._pipeline_config,
                                     output_dir,
                                     outputs))
 
-    if bucket_url:
-      cloud_temp_dir = os.path.join(self._temp_dir, 'cloud')
-      os.mkdir(cloud_temp_dir)
-
-      self._nodes.append(CloudNode(output_dir,
-                                   bucket_url,
-                                   cloud_temp_dir,
-                                   self.is_vod()))
-
-    for node in self._nodes:
-      node.start()
-    return self
-
   def check_status(self) -> ProcessStatus:
     """Checks the status of all the nodes.
 
-    If one node is errored, this returns Errored; otherwise if one node is
-    finished, this returns Finished; this only returns Running if all nodes are
-    running.  If there are no nodes, this returns Finished.
+    If one node is errored, this returns ERRORED; otherwise if one Transcoder, Packager or PeriodConcat 
+    node is running, this returns RUNNING; this only returns FINISHED if all Transcoder, Packager and
+    PeriodConcat nodes are finished. If there are no nodes, this returns FINISHED.
     """
-    if not self._nodes:
-      return ProcessStatus.Finished
+    
+    values: Set[int] = set([ProcessStatus.FINISHED.value])
 
-    value = max(node.check_status().value for node in self._nodes)
+    for node in self._nodes:
+      status: ProcessStatus = node.check_status()
+      if isinstance(node, (TranscoderNode, PackagerNode, PeriodConcatNode)):
+        values.add(status.value)
```

Comment:
`ClouldNode` and also `ExternalCommandNode`, ignoring their status because they do not tell whether the Transcodeing-Packaging pipeline is finished or not.
If all the packager and transcoder nodes are finished then we should know that the required job is done from the controller node, also `CloudNode` is never set to finished, that's another reason i don't check for it.
i think we can give the `CloudNode` a ref to all the nodes as done for the `PeriodConcatNode` to set its status to finished after the last sync.

Reply:
CloudNode is on its way out: #47 

But I don't see any reason to change this method.  It seems that the old rule could still apply: `max(node.check_status().value for node in self._nodes)`.  The only difference is that now, you want the "running" status to overrule "finished".  And you've already changed their values to make that the case with `max()`.

ExternalCommandNode should not be an exception to this rule.  If it's still running, it's still generating input to the other nodes.

CloudNode still requires an exception, at least until we delete it, but that can be handled in CloudNode itself.

Reply:
Here's a suggestion:

1. Pass all packager nodes to both PeriodConcatNode and CloudNode constructors.
2. Add a static method to NodeBase to query the combined status of a list of nodes, based on the `max()` comprehension.
3. Make ControllerNode use the new static method here to compute status of all nodes.
4. PeriodConcatNode can use the new static method to wait for all PackagerNodes to have a Finished status, then combine manifests/playlists, then shut down.
5. If CloudNode sees Finished across PackagerNodes, it can make one additional upload pass, then shut down.